# Neural Network

To predict the 6 analytes using the given spectral data

## Data Preparation

### Load the datasets

In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

In [3]:
data = pd.read_csv('data/data.csv')
target = pd.read_csv('data/target.csv')

### Feature Scaling:

In [4]:
# feature scaling
scaler = StandardScaler()
data = scaler.fit_transform(data)

# convert array back to dataframe
data = pd.DataFrame(data)

In [5]:
data

0         1         2         3         4         5         6     \
0   -0.316769 -0.315116 -0.313438 -0.311752 -0.310101 -0.308489 -0.306886   
1   -0.392352 -0.390931 -0.389515 -0.388140 -0.386799 -0.385506 -0.384272   
2   -0.395732 -0.394205 -0.392682 -0.391210 -0.389802 -0.388441 -0.387181   
3   -0.353691 -0.351974 -0.350242 -0.348509 -0.346777 -0.345091 -0.343411   
4   -0.388814 -0.387189 -0.385506 -0.383811 -0.382188 -0.380599 -0.379026   
..        ...       ...       ...       ...       ...       ...       ...   
176  1.216063  1.212980  1.209800  1.206363  1.202908  1.199597  1.196533   
177  1.405710  1.402492  1.399350  1.396283  1.393269  1.390305  1.387590   
178  1.452745  1.448847  1.445136  1.441636  1.438428  1.435438  1.432754   
179  1.370582  1.367806  1.365097  1.362142  1.359241  1.356519  1.354035   
180  1.325771  1.323243  1.320700  1.317872  1.315010  1.312265  1.309711   

         7         8         9     ...      2790      2791      2792  \
0   -0.305345 -0.303805 -0.302371  ...  0.307467  0.308729  0.309938   
1   -0.383046 -0.381832 -0.380493  ...  0.393738  0.393679  0.393555   
2   -0.385971 -0.384794 -0.383556  ...  0.385905  0.386156  0.386736   
3   -0.341739 -0.340078 -0.338397  ...  0.353261  0.352495  0.351930   
4   -0.377480 -0.375925 -0.374382  ...  0.377174  0.378062  0.379091   
..        ...       ...       ...  ...       ...       ...       ...   
176  1.193830  1.191242  1.188586  ... -1.198214 -1.201348 -1.203690   
177  1.385171  1.382781  1.380179  ... -1.399635 -1.401556 -1.402483   
178  1.430351  1.427990  1.425488  ... -1.458494 -1.459141 -1.458486   
179  1.351736  1.349495  1.346956  ... -1.392243 -1.392203 -1.390443   
180  1.307378  1.305048  1.302498  ... -1.325033 -1.327913 -1.329253   

         2793      2794      2795      2796      2797      2798      2799  
0    0.311335  0.312953  0.314581  0.316054  0.316725  0.316783  0.316769  
1    0.393591  0.393748  0.393820  0.393732  0.393476  0.392953  0.392352  
2    0.387808  0.389403  0.391396  0.393625  0.394835  0.395313  0.395732  
3    0.351762  0.352015  0.352671  0.353454  0.353844  0.353813  0.353691  
4    0.380494  0.382271  0.384355  0.386594  0.387837  0.388371  0.388814  
..        ...       ...       ...       ...       ...       ...       ...  
176 -1.206748 -1.211088 -1.216074 -1.220527 -1.222032 -1.219498 -1.216063  
177 -1.403973 -1.406681 -1.409943 -1.412725 -1.413162 -1.409897 -1.405710  
178 -1.458382 -1.459472 -1.461106 -1.462207 -1.461516 -1.457577 -1.452745  
179 -1.388581 -1.387419 -1.386407 -1.384495 -1.381759 -1.376632 -1.370582  
180 -1.330607 -1.332570 -1.334584 -1.335659 -1.334793 -1.330737 -1.325771  

[181 rows x 2800 columns]

The names of the columns (the wavelengths) are not necessary or useful in prediction, so doesn't matter if renamed.

### Train-Test Split:

In [16]:
# train-test split
X_train, X_test, y_train, y_test = train_test_split(data, target, test_size=0.2, random_state=42)

In [17]:
X_train

0         1         2         3         4         5         6     \
65  -0.294132 -0.292470 -0.290753 -0.289059 -0.287421 -0.285885 -0.284432   
67  -0.284122 -0.282493 -0.280881 -0.279314 -0.277783 -0.276313 -0.274894   
31  -0.306980 -0.305355 -0.303725 -0.302108 -0.300526 -0.299003 -0.297521   
12  -0.283304 -0.281756 -0.280143 -0.278480 -0.276874 -0.275329 -0.273899   
41  -0.276594 -0.275093 -0.273528 -0.271935 -0.270347 -0.268851 -0.267408   
..        ...       ...       ...       ...       ...       ...       ...   
106  1.287983  1.285159  1.282166  1.279110  1.276117  1.273214  1.270348   
14  -0.303077 -0.301640 -0.300190 -0.298733 -0.297363 -0.296033 -0.294755   
92   0.113718  0.112752  0.111822  0.110825  0.109820  0.108743  0.107677   
179  1.370582  1.367806  1.365097  1.362142  1.359241  1.356519  1.354035   
102  1.133316  1.131173  1.128736  1.126225  1.123589  1.120866  1.118060   

         7         8         9     ...      2790      2791      2792  \
65  -0.283031 -0.281643 -0.280298  ...  0.276976  0.278672  0.280488   
67  -0.273518 -0.272146 -0.270753  ...  0.287967  0.288118  0.287882   
31  -0.296112 -0.294748 -0.293489  ...  0.290890  0.292695  0.294566   
12  -0.272542 -0.271209 -0.269888  ...  0.272093  0.273169  0.274347   
41  -0.266029 -0.264725 -0.263381  ...  0.298471  0.296658  0.294318   
..        ...       ...       ...  ...       ...       ...       ...   
106  1.267523  1.264811  1.262283  ... -1.271121 -1.273053 -1.275304   
14  -0.293592 -0.292444 -0.291202  ...  0.288615  0.290797  0.292942   
92   0.106608  0.105618  0.104834  ... -0.096473 -0.099153 -0.101798   
179  1.351736  1.349495  1.346956  ... -1.392243 -1.392203 -1.390443   
102  1.115063  1.112002  1.108749  ... -1.103224 -1.106446 -1.110065   

         2793      2794      2795      2796      2797      2798      2799  
65   0.282665  0.285239  0.288108  0.291088  0.292727  0.293466  0.294132  
67   0.287565  0.287224  0.286867  0.286299  0.285740  0.284973  0.284122  
31   0.296723  0.299148  0.301887  0.304602  0.305979  0.306527  0.306980  
12   0.275795  0.277602  0.279657  0.281763  0.282801  0.283100  0.283304  
41   0.291676  0.288821  0.285722  0.282349  0.280222  0.278439  0.276594  
..        ...       ...       ...       ...       ...       ...       ...  
106 -1.277644 -1.279920 -1.282017 -1.283961 -1.285168 -1.286476 -1.287983  
14   0.295185  0.297425  0.299608  0.301592  0.302569  0.302859  0.303077  
92  -0.104224 -0.106403 -0.108327 -0.110121 -0.111282 -0.112436 -0.113718  
179 -1.388581 -1.387419 -1.386407 -1.384495 -1.381759 -1.376632 -1.370582  
102 -1.113933 -1.117837 -1.121837 -1.126073 -1.128687 -1.130907 -1.133316  

[144 rows x 2800 columns]

In [18]:
X_test

0         1         2         3         4         5         6     \
19  -0.355102 -0.353486 -0.351813 -0.350161 -0.348585 -0.347129 -0.345743   
42  -0.282375 -0.280648 -0.278907 -0.277137 -0.275392 -0.273708 -0.272107   
153  2.635599  2.635815  2.636200  2.636678  2.637186  2.637641  2.637959   
78   0.824260  0.821456  0.818978  0.816870  0.815100  0.813557  0.812046   
145  1.015041  1.018451  1.022076  1.025791  1.029497  1.033103  1.036628   
15  -0.311459 -0.309956 -0.308470 -0.307018 -0.305663 -0.304370 -0.303054   
24  -0.391316 -0.389854 -0.388355 -0.386846 -0.385327 -0.383907 -0.382503   
68  -0.307925 -0.306302 -0.304696 -0.303123 -0.301648 -0.300190 -0.298805   
113 -0.297175 -0.295772 -0.294428 -0.293066 -0.291698 -0.290370 -0.289084   
118  1.324047  1.320891  1.317578  1.314243  1.311018  1.307931  1.304869   
93  -0.043058 -0.042649 -0.042349 -0.042057 -0.041766 -0.041416 -0.041089   
159 -2.130599 -2.136321 -2.142001 -2.147633 -2.153165 -2.158625 -2.164032   
69  -0.302385 -0.300761 -0.299060 -0.297392 -0.295759 -0.294187 -0.292654   
142  1.042602  1.046097  1.049645  1.053181  1.056717  1.060231  1.063713   
45  -0.295512 -0.293861 -0.292196 -0.290555 -0.288949 -0.287445 -0.286004   
16  -0.306920 -0.305128 -0.303342 -0.301641 -0.300006 -0.298441 -0.296947   
51  -0.278903 -0.277137 -0.275327 -0.273530 -0.271788 -0.270129 -0.268531   
125 -0.007011 -0.006555 -0.006221 -0.005950 -0.005694 -0.005384 -0.005109   
96   0.022560  0.022885  0.023109  0.023290  0.023555  0.023941  0.024323   
56  -0.269401 -0.267628 -0.265852 -0.264089 -0.262403 -0.260810 -0.259269   
97   1.252258  1.250115  1.247793  1.245384  1.242902  1.240403  1.237720   
120  1.340220  1.338327  1.336338  1.334465  1.332587  1.330704  1.328788   
143  0.982216  0.985589  0.989027  0.992495  0.995934  0.999294  1.002574   
30  -0.311672 -0.310200 -0.308715 -0.307264 -0.305858 -0.304492 -0.303156   
9   -0.293006 -0.291341 -0.289653 -0.287967 -0.286285 -0.284684 -0.283145   
172  0.885115  0.882563  0.880060  0.877540  0.874920  0.872410  0.870097   
60  -0.300390 -0.298907 -0.297431 -0.295969 -0.294563 -0.293228 -0.291945   
18  -0.315571 -0.313920 -0.312244 -0.310580 -0.308951 -0.307339 -0.305799   
148  1.784769  1.786898  1.788957  1.790983  1.792974  1.794913  1.796791   
173  1.028007  1.025243  1.022645  1.020002  1.017313  1.014726  1.012327   
109 -0.487100 -0.486026 -0.484954 -0.483942 -0.482973 -0.481994 -0.480999   
55  -0.274558 -0.272970 -0.271371 -0.269774 -0.268266 -0.266871 -0.265560   
140 -1.029213 -1.034923 -1.040618 -1.046253 -1.051790 -1.057270 -1.062724   
126 -0.015992 -0.015394 -0.014927 -0.014501 -0.014100 -0.013697 -0.013297   
66  -0.293413 -0.291802 -0.290178 -0.288556 -0.287001 -0.285537 -0.284146   
29  -0.302079 -0.300371 -0.298659 -0.297001 -0.295409 -0.293929 -0.292490   
117  1.352447  1.349433  1.346427  1.343414  1.340438  1.337582  1.334862   

         7         8         9     ...      2790      2791      2792  \
19  -0.344429 -0.343180 -0.341912  ...  0.340416  0.341895  0.343482   
42  -0.270569 -0.269033 -0.267570  ...  0.297368  0.296231  0.294651   
153  2.638116  2.638125  2.638064  ... -2.634756 -2.633414 -2.632869   
78   0.810543  0.809171  0.807972  ... -0.859986 -0.854470 -0.849215   
145  1.040053  1.043411  1.046847  ... -1.017165 -1.016343 -1.015912   
15  -0.301791 -0.300531 -0.299335  ...  0.303902  0.304389  0.304934   
24  -0.381139 -0.379829 -0.378499  ...  0.383391  0.384632  0.385814   
68  -0.297472 -0.296194 -0.294981  ...  0.297898  0.298984  0.300117   
113 -0.287829 -0.286619 -0.285454  ...  0.319660  0.317251  0.314478   
118  1.301836  1.299072  1.296605  ... -1.333095 -1.331216 -1.329523   
93  -0.040826 -0.040622 -0.040554  ...  0.071771  0.067347  0.062971   
159 -2.169349 -2.174434 -2.179333  ...  2.142769  2.144785  2.145523   
69  -0.291132 -0.289632 -0.288218  ...  0.297225  0.298637  0.299680   
142  1.067156  1.070624  1.074210  ... -1.053002 -1.050486 -1.048556   
45

### Dimensionality Reduction:

 Use PCA to reduce the input dimensionality if the dataset is too high-dimensional.

In [40]:
# Dimensionality Reduction
pca = PCA(n_components=0.95)
X_train_pca = pca.fit_transform(X_train)
X_test_pca = pca.transform(X_test)

# convert back to a dataframe
X_train_pca = pd.DataFrame(X_train_pca)
X_test_pca = pd.DataFrame(X_test_pca)

In [41]:
X_train_pca.shape

(144, 3)

In [42]:
X_train_pca

0          1          2
0   -30.397715  -7.173837   0.505082
1   -30.639217  -7.951355   1.682380
2   -30.803980  -6.855545   0.180291
3   -30.928344  -7.593637   0.553239
4   -29.201630  -9.662276   1.238003
..         ...        ...        ...
139  73.491355  -0.817233 -12.861807
140 -30.279794  -7.323924   0.720564
141   7.762607   4.279655 -14.742861
142  92.428913 -13.540680  15.012772
143  69.878793   2.200257 -13.313499

[144 rows x 3 columns]

In [43]:
X_test_pca.shape

(37, 3)

In [44]:
X_test_pca

0          1          2
0  -31.353769  -5.490578  -0.029496
1  -29.071803  -9.115020   0.351931
2   42.753934 -90.722974  21.305769
3   36.982381  -8.221943   3.612671
4  -34.085430 -74.027118  11.974625
5  -30.136958  -7.429598   1.266313
6  -30.365386  -4.896786   0.513040
7  -30.055368  -7.852576   2.747932
8  -28.149508  -4.202456  -6.073699
9   76.569013   0.869426 -11.531683
10  -9.851589   2.798545 -19.612455
11  27.850689  94.218495  14.541010
12 -30.350933  -7.537003   1.873046
13 -35.359282 -76.491825  16.945608
14 -27.120806 -10.191789   0.899216
15 -30.248057  -7.643882   1.415930
16 -31.501428  -7.820246   0.180954
17 -14.918157   1.019441 -17.715787
18 -11.239467  -0.653864 -15.826107
19 -31.147900  -8.383498   0.380904
20  70.816697  -1.887904 -10.057136
21  67.309983  -2.203630 -12.803268
22 -34.116016 -73.958465  13.895607
23 -30.754469  -6.671349   0.159098
24 -30.729184  -7.435355   0.786364
25  62.875315  -7.279404  13.071894
26 -30.330610  -7.257699   0.702954
27 -30.075954  -7.481759   1.108389
28   3.008272 -80.742791  15.989256
29  71.909598  -8.162476  13.692019
30 -23.313785  -1.401937  -4.728797
31 -31.064194  -8.409400   0.732484
32  48.985367  66.094734  14.451243
33 -14.496401   1.462176 -17.982878
34 -30.365823  -7.571352   1.276961
35 -31.052657  -6.628127  -0.556607
36  74.515615   1.187770 -14.670252

---

# Define the Neural Network Architecture

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

# Define the model
model = Sequential()

# Input layer
model.add(Dense(units=128, input_dim=X_train_pca.shape[1], activation='relu'))

# Hidden layers
model.add(Dense(units=64, activation='relu'))
model.add(Dropout(rate=0.3))
model.add(Dense(units=32, activation='relu'))
model.add(Dropout(rate=0.3))

# Output layer
model.add(Dense(units=6, activation='linear'))

# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae'])

# Summary of the model
model.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_12 (Dense)            (None, 128)               512       
                                                                 
 dense_13 (Dense)            (None, 64)                8256      
                                                                 
 dropout_6 (Dropout)         (None, 64)                0         
                                                                 
 dense_14 (Dense)            (None, 32)                2080      
                                                                 
 dropout_7 (Dropout)         (None, 32)                0         
                                                                 
 dense_15 (Dense)            (None, 6)                 198       
                                                                 
Total params: 11046 (43.15 KB)
Trainable params: 11046

# Model Training

In [50]:
# validation data for model evaluation split from training data
X_train_pca, X_val_pca, y_train, y_val = train_test_split(X_train_pca, y_train, test_size=0.2, random_state=0)

In [ ]:
# Train the model
history = model.fit(X_train_pca, y_train, epochs=100, batch_size=32, validation_data=(X_val_pca, y_val))

# Test the model
loss, mae = model.evaluate(X_test_pca, y_test)
print(f'Test Loss: {loss:.4f}')
print(f'Test MAE: {mae:.4f}')

Epoch 1/100
4/4 [==============================] - 0s 19ms/step - loss: 32.2818 - mae: 2.6109 - val_loss: 43.8040 - val_mae: 2.9844
Epoch 2/100
4/4 [==============================] - 0s 8ms/step - loss: 31.9390 - mae: 2.5765 - val_loss: 44.7785 - val_mae: 3.0417
Epoch 3/100
4/4 [==============================] - 0s 7ms/step - loss: 30.7405 - mae: 2.5983 - val_loss: 48.0682 - val_mae: 3.0987
Epoch 4/100
4/4 [==============================] - 0s 7ms/step - loss: 33.5641 - mae: 2.5845 - val_loss: 51.3662 - val_mae: 3.0892
Epoch 5/100
4/4 [==============================] - 0s 7ms/step - loss: 35.8498 - mae: 2.7120 - val_loss: 51.1884 - val_mae: 3.0752
Epoch 6/100
4/4 [==============================] - 0s 7ms/step - loss: 25.8877 - mae: 2.4270 - val_loss: 50.4233 - val_mae: 3.0791
Epoch 7/100
4/4 [==============================] - 0s 7ms/step - loss: 28.1375 - mae: 2.5306 - val_loss: 51.1347 - val_mae: 3.1534
Epoch 8/100
4/4 [==============================] - 0s 8ms/step - loss: 31.8315 - m

## Model Evaluation

In [58]:
from sklearn.metrics import accuracy_score, r2_score, mean_squared_error, mean_absolute_error

# Calculate metrics

# Predict on test data
y_pred = model.predict(X_test_pca)

r2 = r2_score(y_test, y_pred)
print(f'R^2 Score: {r2:.4f}')

MSE = mean_squared_error(y_test, y_pred)
print(f'MSE: {MSE:.4f}')

MAE = mean_absolute_error(y_test, y_pred)
print(f'MAE: {MAE:.4f}')

2/2 [==============================] - 0s 2ms/step
R^2 Score: 0.3561
MSE: 25.9599
MAE: 2.4431


---

# Predicting the target variables using the complete dataset

### Train - Test Split

#### Validation data

In [62]:
# split training data into training and validation data
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=0)

### Neural Network Architecture

In [63]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout

# Define the model
model = Sequential()

# Input layer
model.add(Dense(units=128, input_dim=X_train.shape[1], activation='relu'))

# Hidden layers
model.add(Dense(units=64, activation='relu'))
model.add(Dropout(rate=0.3))
model.add(Dense(units=32, activation='relu'))
model.add(Dropout(rate=0.3))

# Output layer
model.add(Dense(units=6, activation='linear'))

# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae'])

# Summary of the model
model.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_16 (Dense)            (None, 128)               358528    
                                                                 
 dense_17 (Dense)            (None, 64)                8256      
                                                                 
 dropout_8 (Dropout)         (None, 64)                0         
                                                                 
 dense_18 (Dense)            (None, 32)                2080      
                                                                 
 dropout_9 (Dropout)         (None, 32)                0         
                                                                 
 dense_19 (Dense)            (None, 6)                 198       
                                                                 
Total params: 369062 (1.41 MB)
Trainable params: 36906

### Model Training

In [64]:
# Train the model
history = model.fit(X_train, y_train, epochs=100, batch_size=32, validation_data=(X_val, y_val))

# Test the model
loss, mae = model.evaluate(X_test, y_test)

print(f'Test Loss: {loss:.4f}')
print(f'Test MAE: {mae:.4f}')

Epoch 1/100
4/4 [==============================] - 0s 46ms/step - loss: 109.0499 - mae: 4.4334 - val_loss: 91.2992 - val_mae: 4.9005
Epoch 2/100
4/4 [==============================] - 0s 13ms/step - loss: 101.8643 - mae: 5.6191 - val_loss: 86.5615 - val_mae: 4.7888
Epoch 3/100
4/4 [==============================] - 0s 15ms/step - loss: 92.3804 - mae: 5.1437 - val_loss: 91.2639 - val_mae: 4.9505
Epoch 4/100
4/4 [==============================] - 0s 24ms/step - loss: 87.6013 - mae: 4.8773 - val_loss: 83.8811 - val_mae: 4.6287
Epoch 5/100
4/4 [==============================] - 0s 16ms/step - loss: 82.9972 - mae: 4.7677 - val_loss: 77.4590 - val_mae: 4.5588
Epoch 6/100
4/4 [==============================] - 0s 23ms/step - loss: 79.7243 - mae: 4.6571 - val_loss: 77.5201 - val_mae: 4.4384
Epoch 7/100
4/4 [==============================] - 0s 19ms/step - loss: 74.6204 - mae: 4.6049 - val_loss: 74.9706 - val_mae: 4.4891
Epoch 8/100
4/4 [==============================] - 0s 16ms/step - loss: 65

### Model Evaluation

In [66]:
# Calculate metrics

# Predict on test data
y_pred = model.predict(X_test)

r2 = r2_score(y_test, y_pred)
print(f'R^2 Score: {r2:.4f}')

MSE = mean_squared_error(y_test, y_pred)
print(f'MSE: {MSE:.4f}')

MAE = mean_absolute_error(y_test, y_pred)
print(f'MAE: {MAE:.4f}')

2/2 [==============================] - 0s 2ms/step
R^2 Score: 0.2032
MSE: 40.6203
MAE: 2.8175


#### Model Evaluation Results on PCA data:

R^2 Score: 0.3561

MSE: 25.9599

MAE: 2.4431

So, it's better to use data with PCA

---

# Let's predict each target variable separately

## Neural Network Architecture

In [74]:
def NN_model(X_train, y_train, X_val, y_val, X_test, y_test):
    # Define the model
    model = Sequential()

    # Input layer
    model.add(Dense(units=128, input_dim=X_train.shape[1], activation='relu'))

    # Hidden layers
    model.add(Dense(units=64, activation='relu'))
    model.add(Dropout(rate=0.3))
    model.add(Dense(units=32, activation='relu'))
    model.add(Dropout(rate=0.3))

    # Output layer
    model.add(Dense(units=1, activation='linear'))

    # Compile the model
    model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mae'])

    # Train the model
    history = model.fit(X_train, y_train, epochs=100, batch_size=32, validation_data=(X_val, y_val))

    # Test the model
    loss, mae = model.evaluate(X_test, y_test)

    print(f'Test Loss: {loss:.4f}')
    print(f'Test MAE: {mae:.4f}')

    # Calculate metrics

    # Predict on test data
    y_pred = model.predict(X_test)

    r2 = r2_score(y_test, y_pred)
    print(f'R^2 Score: {r2:.4f}')

    MSE = mean_squared_error(y_test, y_pred)
    print(f'MSE: {MSE:.4f}')

    MAE = mean_absolute_error(y_test, y_pred)
    print(f'MAE: {MAE:.4f}')

    return model

## We'll do this using 3 variations of the dataset:

1. With the complete dataset

2. With the complete dataset undergone PCA

3. With the selected features dataset for each target variable

---

# 1. With the complete dataset

In [75]:
y_predicted = pd.DataFrame(columns=target.columns)
y_truth = pd.DataFrame(columns=target.columns)

In [76]:
for col in target.columns:
    y = target[col]
    X_train, X_test, y_train, y_test = train_test_split(data, y, test_size=0.2, random_state=0)

    # split training data into training and validation data
    X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=0)

    model = NN_model(X_train, y_train, X_val, y_val, X_test, y_test)

    y_predicted[col] = model.predict(X_test).flatten()
    y_truth[col] = y_test.values

Epoch 1/100
4/4 [==============================] - 1s 61ms/step - loss: 430.8990 - mae: 14.2599 - val_loss: 362.4518 - val_mae: 13.9739
Epoch 2/100
4/4 [==============================] - 0s 13ms/step - loss: 356.1684 - mae: 13.0358 - val_loss: 263.0154 - val_mae: 13.1107
Epoch 3/100
4/4 [==============================] - 0s 11ms/step - loss: 284.8849 - mae: 12.2043 - val_loss: 230.8325 - val_mae: 12.4512
Epoch 4/100
4/4 [==============================] - 0s 11ms/step - loss: 280.9069 - mae: 12.4884 - val_loss: 221.4440 - val_mae: 12.6755
Epoch 5/100
4/4 [==============================] - 0s 10ms/step - loss: 284.7201 - mae: 12.6910 - val_loss: 209.6848 - val_mae: 12.1798
Epoch 6/100
4/4 [==============================] - 0s 10ms/step - loss: 282.9914 - mae: 12.7187 - val_loss: 203.2316 - val_mae: 11.7398
Epoch 7/100
4/4 [==============================] - 0s 10ms/step - loss: 248.1707 - mae: 11.5253 - val_loss: 199.4529 - val_mae: 11.0331
Epoch 8/100
4/4 [==============================]

In [84]:
y_predicted.head()

Glucose (g/L)  Lactate (g/L)  Ethanol (g/L)  Acetate (g/L)  Biomass (g/L)  \
0      20.024462       4.836507       0.581650       0.420489       0.110299   
1      24.768311       0.306895       0.226264       0.303186       0.110299   
2      -0.140375       0.171545       0.144541       0.154827       0.110299   
3      19.868496      33.733025       0.144541       0.342344       2.722680   
4       3.520661       0.101497       0.048329       0.072261       0.180340   

   Formate (g/L)  
0       0.586327  
1       0.338433  
2       0.183050  
3       0.109338  
4       0.049769

In [83]:
y_truth.head()

Glucose (g/L)  Lactate (g/L)  Ethanol (g/L)  Acetate (g/L)  Biomass (g/L)  \
0      30.000000      25.000000            0.5            2.0        0.00000   
1      50.000000       0.000000            0.0            0.0        0.00000   
2       0.000000       0.000000            0.0            1.5        0.00000   
3      34.739658      40.735563            0.0            0.0        3.91524   
4       8.660000       2.180000            0.0            0.0        0.72000   

   Formate (g/L)  
0            1.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0

In [80]:
# r2 score for all target variables
r2 = r2_score(y_truth, y_predicted)
r2

0.42913293043177275

In [82]:
# r2 score for each target variable
r2 = r2_score(y_truth, y_predicted, multioutput='raw_values')

# convert to a dataframe
r2 = pd.DataFrame(r2, index=target.columns, columns=['R^2 Score'])
r2

R^2 Score
Glucose (g/L)   0.549681
Lactate (g/L)   0.568295
Ethanol (g/L)   0.156038
Acetate (g/L)   0.144207
Biomass (g/L)   0.857244
Formate (g/L)   0.299333

### The combined r2 score is: **0.42913293043177275** [curr]

The combined r2 score is: **0.35816286323129254** *[2]*

The combined r2 score is: **0.2843934946185081** *[3]*

The combined r2 score is: **0.33809704738378416** *[4]*


Which is an average score.

Individually for each target variable, the r2 score has high variability. With Biomass being predicted most accurately with a score of 0.857244, and acetate being predicted most poorly with a score of 0.144207.

---

# 2. With the complete dataset undergone PCA

In [85]:
y_predicted = pd.DataFrame(columns=target.columns)
y_truth = pd.DataFrame(columns=target.columns)

In [87]:
for col in target.columns:
    y = target[col]
    X_train, X_test, y_train, y_test = train_test_split(data, y, test_size=0.2, random_state=0)

    # Dimensionality Reduction
    pca = PCA(n_components=0.95)
    X_train_pca = pca.fit_transform(X_train)
    X_test_pca = pca.transform(X_test)

    # convert back to a dataframe
    X_train_pca = pd.DataFrame(X_train_pca)
    X_test_pca = pd.DataFrame(X_test_pca)

    # split training data into training and validation data
    X_train, X_val, y_train, y_val = train_test_split(X_train_pca, y_train, test_size=0.2, random_state=0)

    model = NN_model(X_train, y_train, X_val, y_val, X_test_pca, y_test)

    y_predicted[col] = model.predict(X_test_pca).flatten()
    y_truth[col] = y_test.values

Epoch 1/100
4/4 [==============================] - 0s 35ms/step - loss: 392.6013 - mae: 13.4916 - val_loss: 332.9871 - val_mae: 14.1526
Epoch 2/100
4/4 [==============================] - 0s 8ms/step - loss: 315.3870 - mae: 11.9153 - val_loss: 275.4960 - val_mae: 13.5173
Epoch 3/100
4/4 [==============================] - 0s 7ms/step - loss: 300.3630 - mae: 12.7339 - val_loss: 252.5392 - val_mae: 13.2818
Epoch 4/100
4/4 [==============================] - 0s 8ms/step - loss: 269.3015 - mae: 12.1427 - val_loss: 244.6939 - val_mae: 12.9784
Epoch 5/100
4/4 [==============================] - 0s 7ms/step - loss: 285.2235 - mae: 12.8012 - val_loss: 240.3986 - val_mae: 12.8313
Epoch 6/100
4/4 [==============================] - 0s 7ms/step - loss: 286.8362 - mae: 12.8263 - val_loss: 231.1400 - val_mae: 12.7048
Epoch 7/100
4/4 [==============================] - 0s 7ms/step - loss: 228.8675 - mae: 11.1457 - val_loss: 224.3524 - val_mae: 12.6588
Epoch 8/100
4/4 [==============================] - 0s 

In [88]:
y_predicted.head()

Glucose (g/L)  Lactate (g/L)  Ethanol (g/L)  Acetate (g/L)  Biomass (g/L)  \
0      16.823990       3.329969       0.414321       0.308510       0.066362   
1       6.614796      -0.002380       0.108477       0.093146       0.051840   
2       2.313177       0.181588       0.114306       0.090248       0.062036   
3      22.379244      31.750568       0.108698       0.034434       2.765123   
4       5.483399       0.295539       0.143515       0.050766       0.284747   

   Formate (g/L)  
0       0.216090  
1       0.093849  
2       0.109636  
3       0.156546  
4       0.075028

In [89]:
y_truth.head()

Glucose (g/L)  Lactate (g/L)  Ethanol (g/L)  Acetate (g/L)  Biomass (g/L)  \
0      30.000000      25.000000            0.5            2.0        0.00000   
1      50.000000       0.000000            0.0            0.0        0.00000   
2       0.000000       0.000000            0.0            1.5        0.00000   
3      34.739658      40.735563            0.0            0.0        3.91524   
4       8.660000       2.180000            0.0            0.0        0.72000   

   Formate (g/L)  
0            1.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0

In [90]:
# r2 score for all target variables
r2 = r2_score(y_truth, y_predicted)
r2

0.35816286323129254

In [91]:
# r2 score for each target variable
r2 = r2_score(y_truth, y_predicted, multioutput='raw_values')

# convert to a dataframe
r2 = pd.DataFrame(r2, index=target.columns, columns=['R^2 Score'])
r2

R^2 Score
Glucose (g/L)   0.057638
Lactate (g/L)   0.604816
Ethanol (g/L)   0.128175
Acetate (g/L)   0.165103
Biomass (g/L)   0.884916
Formate (g/L)   0.308329

The combined r2 score is: **0.42913293043177275** [1]

### The combined r2 score is: **0.35816286323129254** [curr]

The combined r2 score is: **0.2843934946185081** [3]

The combined r2 score is: **0.33809704738378416** [4]


Which is an average score.

Individually for each target variable, the r2 score has high variability. With Biomass being predicted most accurately with a score of 0.884916 (higher than before) and Glucose being predicted most poorly with a score of 0.057638.

---

# 3. With the selected features dataset for each target variable

In [102]:
y_predicted = pd.DataFrame(columns=target.columns)
y_truth = pd.DataFrame(columns=target.columns)

In [103]:
glucose = pd.read_csv('modified_data/data_glucose2.csv')
lactate = pd.read_csv('modified_data/data_lactate.csv')
ethanol = pd.read_csv('modified_data/data_ethanol.csv')
acetate = pd.read_csv('modified_data/data_acetate.csv')
biomass = pd.read_csv('modified_data/data_biomass.csv')
formate = pd.read_csv('modified_data/data_formate.csv')

In [104]:
datasets = [glucose, lactate, ethanol, acetate, biomass, formate]

In [105]:
for dataset in datasets:
    # y is the last column in the dataset
    y = dataset.iloc[:, -1]
    X = dataset.iloc[:, :-1]

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

    # split training data into training and validation data
    X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.2, random_state=0)

    model = NN_model(X_train, y_train, X_val, y_val, X_test, y_test)

    y_predicted[dataset.columns[-1]] = model.predict(X_test).flatten()
    y_truth[dataset.columns[-1]] = y_test.values

Epoch 1/100
4/4 [==============================] - 0s 40ms/step - loss: 408.7224 - mae: 13.1298 - val_loss: 442.9629 - val_mae: 14.7049
Epoch 2/100
4/4 [==============================] - 0s 8ms/step - loss: 404.9343 - mae: 13.0881 - val_loss: 438.7206 - val_mae: 14.6806
Epoch 3/100
4/4 [==============================] - 0s 8ms/step - loss: 401.1028 - mae: 13.0240 - val_loss: 433.7328 - val_mae: 14.6608
Epoch 4/100
4/4 [==============================] - 0s 8ms/step - loss: 397.4810 - mae: 13.0408 - val_loss: 427.7477 - val_mae: 14.6414
Epoch 5/100
4/4 [==============================] - 0s 8ms/step - loss: 389.8109 - mae: 13.0172 - val_loss: 420.2066 - val_mae: 14.6186
Epoch 6/100
4/4 [==============================] - 0s 9ms/step - loss: 380.6319 - mae: 12.9170 - val_loss: 410.9977 - val_mae: 14.5888
Epoch 7/100
4/4 [==============================] - 0s 8ms/step - loss: 376.1442 - mae: 12.8747 - val_loss: 399.6980 - val_mae: 14.5521
Epoch 8/100
4/4 [==============================] - 0s 

In [106]:
y_predicted.head()

Glucose (g/L)  Lactate (g/L)  Ethanol (g/L)  Acetate (g/L)  Biomass (g/L)  \
0      12.044641       5.814450       0.337607       0.181046       0.139045   
1       8.300912       5.204381       0.208007       0.117998       0.127974   
2       4.568072       5.015829       0.127050       0.089438       0.165664   
3      19.185236      15.855120       0.501427       0.325027       3.737548   
4       3.817746       3.584099       0.064924       0.037689       0.362659   

   Formate (g/L)  
0       0.216647  
1       0.154979  
2       0.138042  
3       0.416546  
4       0.129893

In [107]:
y_truth.head()

Glucose (g/L)  Lactate (g/L)  Ethanol (g/L)  Acetate (g/L)  Biomass (g/L)  \
0      30.000000      25.000000            0.5            2.0        0.00000   
1      50.000000       0.000000            0.0            0.0        0.00000   
2       0.000000       0.000000            0.0            1.5        0.00000   
3      34.739658      40.735563            0.0            0.0        3.91524   
4       8.660000       2.180000            0.0            0.0        0.72000   

   Formate (g/L)  
0            1.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0

In [108]:
# r2 score for all target variables
r2 = r2_score(y_truth, y_predicted)
r2

0.2843934946185081

In [109]:
# r2 score for each target variable
r2 = r2_score(y_truth, y_predicted, multioutput='raw_values')

# convert to a dataframe
r2 = pd.DataFrame(r2, index=target.columns, columns=['R^2 Score'])
r2

R^2 Score
Glucose (g/L)   0.052366
Lactate (g/L)   0.161480
Ethanol (g/L)   0.079092
Acetate (g/L)   0.117112
Biomass (g/L)   0.970402
Formate (g/L)   0.325907

The combined r2 score is: **0.42913293043177275** [1]

The combined r2 score is: **0.35816286323129254** [2]

### The combined r2 score is: **0.2843934946185081** [curr] THE WORST

The combined r2 score is: **0.33809704738378416** [4]

Which is a below average score. :(

BUT

Individually for each target variable, the r2 score has high variability. With Biomass being predicted most accurately with a score of ***0.970402***, and acetate being predicted most poorly with a score of 0.052366.

---

# 4. With the selected features dataset (after dimensionality reduction by PCA) for each target variable

In [93]:
y_predicted = pd.DataFrame(columns=target.columns)
y_truth = pd.DataFrame(columns=target.columns)

In [94]:
glucose = pd.read_csv('modified_data/data_glucose2.csv')
lactate = pd.read_csv('modified_data/data_lactate.csv')
ethanol = pd.read_csv('modified_data/data_ethanol.csv')
acetate = pd.read_csv('modified_data/data_acetate.csv')
biomass = pd.read_csv('modified_data/data_biomass.csv')
formate = pd.read_csv('modified_data/data_formate.csv')

In [95]:
datasets = [glucose, lactate, ethanol, acetate, biomass, formate]

In [97]:
for dataset in datasets:
    # y is the last column in the dataset
    y = dataset.iloc[:, -1]
    X = dataset.iloc[:, :-1]

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

    # Dimensionality Reduction
    pca = PCA(n_components=0.95)
    X_train_pca = pca.fit_transform(X_train)
    X_test_pca = pca.transform(X_test)

    # convert back to a dataframe
    X_train_pca = pd.DataFrame(X_train_pca)
    X_test_pca = pd.DataFrame(X_test_pca)

    # split training data into training and validation data
    X_train, X_val, y_train, y_val = train_test_split(X_train_pca, y_train, test_size=0.2, random_state=0)

    model = NN_model(X_train, y_train, X_val, y_val, X_test_pca, y_test)

    y_predicted[dataset.columns[-1]] = model.predict(X_test_pca).flatten()
    y_truth[dataset.columns[-1]] = y_test.values

Epoch 1/100
4/4 [==============================] - 0s 34ms/step - loss: 409.5983 - mae: 13.1372 - val_loss: 446.0562 - val_mae: 14.7243
Epoch 2/100
4/4 [==============================] - 0s 7ms/step - loss: 407.7460 - mae: 13.1158 - val_loss: 443.8323 - val_mae: 14.7134
Epoch 3/100
4/4 [==============================] - 0s 8ms/step - loss: 405.9049 - mae: 13.0955 - val_loss: 441.0459 - val_mae: 14.7002
Epoch 4/100
4/4 [==============================] - 0s 8ms/step - loss: 403.3758 - mae: 13.0840 - val_loss: 437.7345 - val_mae: 14.6847
Epoch 5/100
4/4 [==============================] - 0s 8ms/step - loss: 400.2840 - mae: 13.0742 - val_loss: 433.5514 - val_mae: 14.6682
Epoch 6/100
4/4 [==============================] - 0s 8ms/step - loss: 396.8241 - mae: 13.0302 - val_loss: 428.2809 - val_mae: 14.6517
Epoch 7/100
4/4 [==============================] - 0s 7ms/step - loss: 392.4183 - mae: 12.9936 - val_loss: 421.6544 - val_mae: 14.6307
Epoch 8/100
4/4 [==============================] - 0s 

In [98]:
y_predicted.head()

Glucose (g/L)  Lactate (g/L)  Ethanol (g/L)  Acetate (g/L)  Biomass (g/L)  \
0      12.639973       6.241774       0.941666       0.830453       0.025640   
1       8.473408       5.801205       0.087066       0.071283       0.021768   
2       4.326695       6.053002       0.077514       0.087387       0.025963   
3      19.522985      18.556988       0.031337       0.012789       3.682764   
4       4.298193       2.921015       0.030391       0.047471       0.881406   

   Formate (g/L)  
0       0.414972  
1       0.034079  
2       0.231419  
3      -0.010527  
4       0.037265

In [99]:
y_truth.head()

Glucose (g/L)  Lactate (g/L)  Ethanol (g/L)  Acetate (g/L)  Biomass (g/L)  \
0      30.000000      25.000000            0.5            2.0        0.00000   
1      50.000000       0.000000            0.0            0.0        0.00000   
2       0.000000       0.000000            0.0            1.5        0.00000   
3      34.739658      40.735563            0.0            0.0        3.91524   
4       8.660000       2.180000            0.0            0.0        0.72000   

   Formate (g/L)  
0            1.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0

In [100]:
# r2 score for all target variables
r2 = r2_score(y_truth, y_predicted)
r2

0.33809704738378416

In [101]:
# r2 score for each target variable
r2 = r2_score(y_truth, y_predicted, multioutput='raw_values')

# convert to a dataframe
r2 = pd.DataFrame(r2, index=target.columns, columns=['R^2 Score'])
r2

R^2 Score
Glucose (g/L)   0.069825
Lactate (g/L)   0.182412
Ethanol (g/L)   0.157033
Acetate (g/L)   0.253995
Biomass (g/L)   0.993745
Formate (g/L)   0.371572

The combined r2 score is: **0.42913293043177275** [1]

The combined r2 score is: **0.35816286323129254** [2]

The combined r2 score is: **0.2843934946185081** [3]

### The combined r2 score is: **0.33809704738378416** [curr]

Which is an average score. Worse than previous models.

BUT

Individually for each target variable, the r2 score has high variability. With Biomass being predicted most accurately with a score of ***0.993745***, and acetate being predicted most poorly with a score of 0.069825.

---